tests/unit/src/test/scala/tests/UserConfigurationSuite.scala

```diff
@@ -0,0 +1,127 @@
+package tests
+
+import com.google.gson.JsonParser
+import java.util.Properties
+import scala.meta.internal.metals.UserConfiguration
+import scala.collection.JavaConverters._
+
+object UserConfigurationSuite extends BaseSuite {
+  def check(
+      name: String,
+      original: String,
+      props: Map[String, String] = Map.empty
+  )(fn: Either[List[String], UserConfiguration] => Unit): Unit = {
+    test(name) {
+      val json = new JsonParser().parse(original).getAsJsonObject
+      val jprops = new Properties()
+      jprops.putAll(props.asJava)
+      val obtained = UserConfiguration.fromJson(json, jprops)
+      fn(obtained)
+    }
+  }
+
+  def checkOK(
+      name: String,
+      original: String,
+      props: Map[String, String] = Map.empty
+  )(fn: UserConfiguration => Unit): Unit = {
+    check(name, original, props) {
+      case Left(errs) =>
+        fail(s"Expected success. Obtained error: $errs")
+      case Right(obtained) =>
+        fn(obtained)
+    }
+  }
+  def checkError(
+      name: String,
+      original: String,
+      expected: String
+  ): Unit = {
+    check(name, original) {
+      case Right(ok) =>
+        fail(s"Expected error. Obtained successful value $ok")
+      case Left(errs) =>
+        val obtained = errs.mkString("\n")
+        assertNoDiff(obtained, expected)
+    }
+  }
+
+  checkOK(
+    "basic",
+    """
+      |{
+      | "sbt-launcher": "launch",
+      | "java-home": "home",
+      | "sbt-options": "a b"
+      |}
+    """.stripMargin
```

Comment:
See my comment above, the client actually sends these nested inside a `"metals"` key.

Reply:
Adding more details to this.

I strongly suspect there's no agreement on how clients will send this.
I'm sure VSCode will wrap the settings in `metals`, but by reading the code I don't think Atom does the same. It looks like it will send them wrapped in `settings`.

@laughedelic I don't have Atom on my machine. If you have a minute, can you confirm which format Atom uses for configuration?

Reply:
Atom seems to send the same structure as vscode
```json
{
  "jsonrpc": "2.0",
  "method": "workspace/didChangeConfiguration",
  "params": {
    "settings": {
      "metals": {
        "version": "0.4.0-SNAPSHOT",
        "javaArgs": [
          "-XX:+UseG1GC",
          "-XX:+UseStringDeduplication",
          "-Xss4m",
          "-Xms1G",
          "-Xmx4G"
        ]
      }
    }
  }
}
```